scheds/rust/scx_layered/src/bpf/main.bpf.c

```diff
@@ -494,14 +524,27 @@ void BPF_STRUCT_OPS(layered_enqueue, struct task_struct *p, u64 enq_flags)
 		preempt_cursor = (cpu + 1) % nr_possible_cpus;
 
 		lstat_inc(LSTAT_PREEMPT, layer, cctx);
-		break;
+		return;
 	}
+
+	lstat_inc(LSTAT_PREEMPT_FAIL, layer, cctx);
 }
 
 void BPF_STRUCT_OPS(layered_dispatch, s32 cpu, struct task_struct *prev)
 {
+	s32 sib = sibling_cpu(cpu);
+	struct cpu_ctx *cctx, *sib_cctx;
 	int idx;
 
+	if (!(cctx = lookup_cpu_ctx(-1)))
+		return;
+
+	if (sib >= 0 && (sib_cctx = lookup_cpu_ctx(sib)) &&
+	    sib_cctx->current_exclusive) {
+		gstat_inc(GSTAT_EXCL_IDLE, cctx);
+		return;
+	}
```

Comment:
Isn't this all racy? Does it just not matter or is there some synchronization I'm missing

Reply:
It is very racy. The mechanism doesn't provide any hard guarantee and just provides best-effort optimization, so should hopefully be okay? It does change the scheduling behavior substantially, so it looks like the races aren't hurting too much. I'll add a comment.